### PR TITLE
Add GetAllChildEntries to get all entries. Remove duplicated code (fi…

### PR DESCRIPTION
**Pull request by chromicant**
…ndMatchingGroup). Refactor GetChildEntries to add everyEntry to list entries that are missing a URL.

Basically added a new endpont, and kept the behavior the same for the other endpoints. GetChildEntries gets called from a few places, but the behavior should be the same.

### CODE REVIEW

File: KeePassRPC/KeePassRPCService.cs [hunk 1]
```diff
@@ -1649,9 +1669,9 @@ private Group[] GetChildGroups(PwDatabase pwd, PwGroup group, bool complete, boo
                 {
                     kpg.ChildGroups = GetChildGroups(pwd, pwg, true, fullDetails);
                     if (fullDetails)
-                        kpg.ChildEntries = (Entry[])GetChildEntries(pwd, pwg, fullDetails);
+                        kpg.ChildEntries = (Entry[])GetChildEntries(pwd, pwg, fullDetails, false);
                     else
-                        kpg.ChildLightEntries = GetChildEntries(pwd, pwg, fullDetails);
+                    	kpg.ChildLightEntries = GetChildEntries(pwd, pwg, fullDetails, false);
```
Comment (luckyrat) [inline KeePassRPC/KeePassRPCService.cs]: whitespace damage here and in a few other places in the diff (don't replace spaces with tabs - let's try to keep things consistent with current style)

File: KeePassRPC/KeePassRPCService.cs [hunk 2]
```diff
@@ -1530,17 +1562,17 @@ public Entry[] GetChildEntries(string uuid)
...
-
+        
         /// <summary>
         /// Returns a list of every entry contained within a group (not recursive)
         /// </summary>
         /// <param name="uuid">the unique ID of the group we're interested in.</param>
         /// <param name="current__"></param>
         /// <returns>the list of every entry directly inside the group.</returns>
-        private LightEntry[] GetChildEntries(PwDatabase pwd, PwGroup group, bool fullDetails)
+        private LightEntry[] GetChildEntries(PwDatabase pwd, PwGroup group, bool fullDetails, bool everyEntry)
```
Comment (luckyrat) [inline KeePassRPC/KeePassRPCService.cs]: Would be clearer to flip the bool and use `urlRequired` instead - `everyEntry` would just leave people wondering which entries would be omitted if they pass in false. It's not ideal that we have to stick with the concept of "Entries" and "AllEntries" on the public API but we can at least make it clear in the private code.

File: KeePassRPC/KeePassRPCService.cs [hunk 3]
```diff
@@ -1556,7 +1588,7 @@ private LightEntry[] GetChildEntries(PwDatabase pwd, PwGroup group, bool fullDet
                     if (EntryIsInRecycleBin(pwe, pwd))
                         continue; // ignore if it's in the recycle bin
 
-                    if (string.IsNullOrEmpty(pwe.Strings.ReadSafe("URL")))
+                    if (string.IsNullOrEmpty(pwe.Strings.ReadSafe("URL")) && !everyEntry)
```
Comment (luckyrat) [inline KeePassRPC/KeePassRPCService.cs]: This would be (fractionally!) faster if you put the boolean test before the string test.

**Review by luckyrat — CHANGES_REQUESTED** [review]
Looks good. I've put a few comments and yes it would be good if you could do some more thorough testing before we merge it into master but this looks like the right direction.

File: KeePassRPC/KeePassRPCService.cs (hunk 2 above)
Comment (chromicant) [reply KeePassRPC/KeePassRPCService.cs]: Yes. I looked into Jayrock and optional parameters - and it appears to be a mess to implement. Far simpler to expand the API, sadly.

File: KeePassRPC/KeePassRPCService.cs (hunk 1 above)
Comment (chromicant) [reply KeePassRPC/KeePassRPCService.cs]: Oops, I'll blame SharpDevelop for that. I also found a lot of trailing whitespace that also got cleaned up as well.

**Comment by chromicant** [comment]
Made the recommended changes. Did notice that there was another place where URL is checked to see if it's null or empty in GetAllLogins. Added the GetAllEntries endpoint to fix that.

My testbed is a forked Go project (https://github.com/chromicant/gokpass - devel branch) which is now talking with the KeePassRPC on Windows. Need to test it on Linux/Mono.

**Comment by luckyrat** [comment]
@chromicant I've incorporated your suggested changes into KeePassRPC v1.8. GitHub wouldn't let me work with this PR so I'm closing it but you should find that any version of KeePassRPC which offers the feature flag "KPRPC_ENTRIES_WITH_NO_URL" will enable you call the GetAllEntries or GetAllChildEntries JSONRPC endpoints.

You can either pull and build the current master branch or keep an eye on the releases page or the update notifications in KeePass for when version 1.8 is released in around a week.

If you try it before the release and find any problems feel free to issue a new PR - it might be many months (or a year) before there's a strong need for the next new version of KeePassRPC so it would be good to get this done right this time around.

**Pull request closed without merging**